### PR TITLE
Add patch for Android Q

### PR DESCRIPTION
**Pull request by gudenau**
I have not tested the code out, but it builds as it is. (Emulator refusing to run for some reason)

### CODE REVIEW

**Comment by gudenau** [comment]
Screenshot_20191025-174407
It seems to work, now that I have a device I can test on.

File: patches/android_frameworks_base-Q.patch [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
++                    && p.mAppMetaData != null) {
++                String sig = p.mAppMetaData.getString("fake-signature");
++                if (sig != null) {
++                    pi.signatures = new Signature[] {new Signature(sig)};
++                }
++            }
++        } catch (Throwable t) {
++            // We should never die because of any failures, this is system code!
++            Log.w("PackageManagerService.FAKE_PACKAGE_SIGNATURE", t);
++	    }
```
Comment (benwaffle) [inline patches/android_frameworks_base-Q.patch]: Could you fix the indentation here?

File: patches/android_frameworks_base-Q.patch [hunk 2]
```diff
@@ -0,0 +1,88 @@
+diff --git a/core/res/AndroidManifest.xml b/core/res/AndroidManifest.xml
+index a3fccd96f22..383c702955f 100644
+--- a/core/res/AndroidManifest.xml
++++ b/core/res/AndroidManifest.xml
+@@ -2628,6 +2628,13 @@
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
@@ -2631,6 +2631,13 @@
```

File: patches/android_frameworks_base-Q.patch [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
++        android:description="@string/permdesc_fakePackageSignature" />
++
+     <!-- @deprecated No longer useful, see
+          {@link android.content.pm.PackageManager#addPackageToPreferred}
+          for details. -->
+diff --git a/core/res/res/values/config.xml b/core/res/res/values/config.xml
+index 0391530c483..5df70cc1bd1 100644
+--- a/core/res/res/values/config.xml
++++ b/core/res/res/values/config.xml
+@@ -1855,6 +1855,8 @@
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
@@ -1871,6 +1871,8 @@
```

File: patches/android_frameworks_base-Q.patch [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
++        <!-- The (faked) microg fused location provider (a free reimplementation) -->
++        <item>com.google.android.gms</item>
+     </string-array>
+ 
+     <!-- This string array can be overriden to enable test location providers initially. -->
+diff --git a/core/res/res/values/strings.xml b/core/res/res/values/strings.xml
+index 152b131af8f..f8d87a37064 100644
+--- a/core/res/res/values/strings.xml
++++ b/core/res/res/values/strings.xml
+@@ -827,6 +827,11 @@
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
@@ -831,6 +831,11 @@
```

File: patches/android_frameworks_base-Q.patch [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
```
delete

File: patches/android_frameworks_base-Q.patch [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+     <!--  Permissions -->
+ 
++    <!-- Title of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
++    <string name="permlab_fakePackageSignature">Spoof package signature</string>
++    <!-- Description of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
++    <string name="permdesc_fakePackageSignature">Allows the app to pretend to be a different app. Malicious applications might be able to use this to access private application data. Legitimate uses include an emulator pretending to be what it emulates. Grant this permission with caution only!</string>
++
+     <!-- Title of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
+     <string name="permlab_statusBar">disable or modify status bar</string>
+     <!-- Description of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
     <!-- Description of an application permission, listed so the user can choose whether they want to allow the application to do this. -->
     <string name="permdesc_statusBar">Allows the app to disable the status bar or add and remove system icons.</string>
```

File: patches/android_frameworks_base-Q.patch [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
+index 7469e099421..48e78eff9db 100644
+--- a/services/core/java/com/android/server/pm/PackageManagerService.java
++++ b/services/core/java/com/android/server/pm/PackageManagerService.java
+@@ -4092,8 +4092,9 @@ public class PackageManagerService extends IPackageManager.Stub
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
@@ -4108,8 +4108,9 @@ public class PackageManagerService extends IPackageManager.Stub
```

File: patches/android_frameworks_base-Q.patch [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ 
+-            PackageInfo packageInfo = PackageParser.generatePackageInfo(p, gids, flags,
+-                    ps.firstInstallTime, ps.lastUpdateTime, permissions, state, userId);
++            PackageInfo packageInfo = mayFakeSignature(p, PackageParser.generatePackageInfo(p, gids, flags,
++                    ps.firstInstallTime, ps.lastUpdateTime, permissions, state, userId),
++                    permissions);
+ 
+             if (packageInfo == null) {
+                 return null;
+@@ -4129,6 +4130,24 @@ public class PackageManagerService extends IPackageManager.Stub
```
Comment (chris42) [inline patches/android_frameworks_base-Q.patch]:
```suggestion
@@ -4145,6 +4146,24 @@ public class PackageManagerService extends IPackageManager.Stub
```

**Comment by memiks** [comment]
Dear Team, 

Could you please add this patch ? I would like to use it on GloDroid image in order to test MicroG with PinePhone.

BR,
Memiks.

**Comment by JonnyTech** [comment]
@memiks if your system image is built with signature spoofing enabled then this patch is not required and you just need to `adb shell pm grant com.android.vending android.permission.FAKE_PACKAGE_SIGNATURE`

**Comment by memiks** [comment]
> 
> 
> @memiks if your system image is built with signature spoofing enabled then this patch is not required and you just need to `adb shell pm grant com.android.vending android.permission.FAKE_PACKAGE_SIGNATURE`

Yes of course, sure, but GloDroid not include the signature spoofing for the moment and this project is based on Android Q https://github.com/glodroid/glodroid_manifest/

Tanhks.

**Comment by GioKas93** [comment]
how to patch this in frameworks/base ?

**Comment by iot-resister** [comment]
when I  try to build for lineage 17.1, I get

```
error: invalid file path 'frameworks/base/core/res/res/values/config.xml.orig'.
error: invalid file path 'frameworks/base/core/res/res/values/strings.xml.orig'.
```

here is my local manifest

```
<?xml version="1.0" encoding="UTF-8"?>
<manifest>
  <project path="device/lenovo/YTX703F"         name="Quallenauge/android_device_lenovo_YTX703F"        revision="lineage-17.1" />
  <project path="device/lenovo/YTX703L"         name="Quallenauge/android_device_lenovo_YTX703L"        revision="lineage-17.1" />
  <project path="device/lenovo/YTX703-common"   name="Quallenauge/android_device_lenovo_YTX703-common"  revision="lineage-17.1" />
  <project path="kernel/lenovo/msm8976"         name="Quallenauge/android_kernel_lenovo_msm8976"        revision="lineage-17.1" />
  <project path="vendor/lenovo"                 name="Quallenauge/proprietary_vendor_lenovo"            revision="lineage-17.1"/>
  <project name="LineageOS/android_hardware_sony_timekeep" path="hardware/sony/timekeep" remote="github" revision="lineage-17.1" />
</manifest>
```

**Comment by chris42** [comment]
> when I try to build for lineage 17.1, I get
> 
> ```
> error: invalid file path 'frameworks/base/core/res/res/values/config.xml.orig'.
> error: invalid file path 'frameworks/base/core/res/res/values/strings.xml.orig'.
> ```
> 

You probably get a few index errors in the patching itself. patch then creates the orig files. Either correct the indexes (as I did once before above) or delete the .orig files before build.

**Comment by nift4** [comment]
The java part doesn't merge anymore in Lineage because they added some new code directly above it

**Comment by Morelcia** [comment]
Protip for everyone coming here from notification after cherry-picking kdrag0n patch - amend commit and remove mention of this one from description to avoid spamming on every rebase :)

**Comment by mar-v-in** [comment]
Imported to https://github.com/microg/patches

**Pull request closed without merging**